Added a git guide for new members (#54)
NUbots/NUbook: src/book/04-general/01-git.mdx

---
section: Guides
chapter: General
title: Git
description: Contributing to a NUbots project with git.
slug: /guides/general/git
---

## Prerequisites

> This guide assumes you know the basics of git and basic shell commands.

Reviewer: Maybe specify what is basic.

Author: I have added which git commands are prerequisite, but I am not sure what to say about shell. It's mostly that you just can't be shellphobic lol

## Getting Code Into a Repository

> Suppose you have a file you have made which you would like to add to a NUbots repository, called `my_file.cpp`, as part of your project to make the robot stand up.

Reviewer: The wording is a little hard to follow. Maybe something like
```suggestion
Let's suppose you have made a file, which you would like to add to a NUbots repository, called `my_file.cpp`. Suppose this file is part of your project to make the robot stand up.
```

Author: I have made changes along these lines. let me know what u think.

> In terminal, change directory to where you have the repo cloned.

Reviewer: ```suggestion
In the terminal, change directory to where you have the repo cloned.
```

> In the terminal, change directory to where you have the repo cloned.

Reviewer: ```suggestion
In the terminal, change directory to the location where the repository is cloned.
```

### Select Branch

> If you are working within an existing project, then you need to find out which branch the work is on. To list the branches, use `git branch`. Once you can see which branch you to push to swap to it:

Reviewer: ```suggestion
If you are working within an existing project, then you need to find out which branch the work is on. To list the branches, use `git branch`. When you have found the branch you want to push to, swap to it:
```

> If you are working within an existing project, then you need to find out which branch the work is on. To list the branches, use `git branch`. When you have found the branch you want to push to, swap to it:

Reviewer: ```suggestion
If you are working within an existing project, then you need to find out which branch the work is on. To list the branches, use the command `git branch`. When you have found the branch you want to push to, swap to it using the following command:
```

```sh
git checkout <branch name>
```

> If you are not working on an existing project, you need to make a branch to commit your changes to. The format for branch names is `surname/project_name`. So for this example it could be `citizen/stand_up`.

Reviewer: ```suggestion
Where `<branch name>` corresponds to the branch you found in the last step. 

If you are not working on an existing project, you need to make a branch to commit your changes to. The format for branch names is `surname/project_name`. So for this example it could be `citizen/stand_up`, if your surname is Citizen.
```

### Making a Pull Request

Once you have a set of commits on your branch which you think should be added to the main codebase, you can make a Pull Request (PR). PRs are where users ask the repository to pull their branch into the `master` branch.

> Github has documentation about about PRs [here.](#https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/creating-a-pull-request)

Reviewer: ```suggestion
GitHub has documentation about PRs [here.](#https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/creating-a-pull-request)
```

Title your PR with a description of the changes that it will make. For this example we could use `Add stand-up script`. If you need to give more detail about what the code does, put that as a comment.

> To be approved, you should add reviewers to the pull request. If you have a project mentor, they should be a reviewer, as well as someone else who knows about that part of the codebase. Once these people approve the changes, they can be merged in.

Reviewer: ```suggestion
To be approved, you should add reviewers to the pull request. If you have a project mentor, they should be a reviewer, as well as someone else who knows about that part of the codebase. Once the reviewers have approved the changes, the changes can be merged into the master branch.
```

### Code Review

> The reviewers will give constructive feedback about your code, by making comments on code snippets from your changes. You can respond to these on github. They can also directly suggest changes to specific sections. Additional commits you make to the branch will update the PR and will reset their review of changed sections.

Reviewer: ```suggestion
The reviewers will give constructive feedback about your code, by making comments on code snippets from your changes. You can respond to these on GitHub. They can also directly suggest changes to specific sections. Additional commits you make to the branch will update the PR and will reset their review of changed sections.
```

> The reviewers will give constructive feedback about your code, by making comments on code snippets from your changes. You can respond to these on Github. They can also directly suggest changes to specific sections. Additional commits you make to the branch will update the PR and will reset their review of changed sections.

Reviewer: ```suggestion
The reviewers will give constructive feedback about your code, by making comments on code snippets from your changes. You can respond to these on GitHub. They can also directly suggest changes to specific sections. Additional commits you make to the branch will update the PR and will reset reviews of the changed sections.
```

> Once the reviewers are happy with it, and all of the comments are resolved, you can `Squash and Merge`. Squashing removes detailed commit descriptions, so that `master` is not messy with every commit ever. Once the code has been squashed and merged, it is in the codebase and you have succeeded in contributing code to NUbots.

Reviewer: ```suggestion
Once the reviewers have approved the PR, and all of the comments are resolved, you can `Squash and Merge`. Squashing removes detailed commit descriptions, so that `master` is not messy with every commit ever. Once the code has been squashed and merged, it is in the codebase and you have succeeded in contributing code to NUbots.
```

> Once the reviewers are happy with it, and all of the comments are resolved, you can `Squash and Merge`. Squashing removes detailed commit descriptions, so that `master` is not messy with every commit ever. Once the code has been squashed and merged, it is in the codebase and you have succeeded in contributing code to NUbots.

Reviewer: Maybe mention formatting and no errors with buildkite, netlify, or whatever

## Commit Etiquette

Commits should make minor changes, and the changes in each commit should have a common theme. A general mantra for contributing code is `commit often, push once`, but this isn't strict at all. Committing often means that the code is built up incrementally. Pushing once means that you don't spam all the other contributors with notifications.

> ## Formatting and buildkite

Reviewer: ```suggestion
## Formatting and Buildkite
```

> Before making a pull request, it is important that your code adheres to the formatting rules. For branches in the NUbots repo, running `./b format` before your last commit should fix formatting issues. Code in the NUbook repo must be formatted with `yarn format` before it can be committed.

Reviewer: Is there a similar thing for NUsight2 @JosephusPaye?

Reviewer 2: For NUsight, it's `yarn eslint` to find problems and `yarn eslint:fix` to fix.

If your code is in a different repository and it uses C++ or protobuf files, use the `.clang-format` file in the NUbots repo to format them.

> If you get a message that your build failed, try running the relevant format commands and then committing the changes which they made. You won't be able to merge the PR until the build checks pass.

Reviewer: ```suggestion
If you get a message that your build failed, try running the relevant format commands and then committing any format changes. You won't be able to merge the PR until the build checks pass.
```

Reviewer: Not sure on this one

## Work in Progress PRs

> If you have code which isn't done yet, but which you would like feedback on, you can make a draft PR to get your code reviewed. Prepend `[WIP]` to the PR title when making draft pull requests. Draft PRs cannot be merged, so there is no need to worry about merging in bad code. Github has documentation about draft PRs [here.](#https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/about-pull-requests#draft-pull-requests)

Reviewer: there is a button you can press for this (drop down on 'create pull request' then change to draft pull request)

Author: I was steering clear of explicitly saying which button and where the button is, but we can add if you reckon

Reviewer: Ye fair enough and you link to the GitHub page about it anyway

> If you have code which isn't done yet, but which you would like feedback on, you can make a draft PR to get your code reviewed. Prepend `[WIP]` to the PR title when making draft pull requests. Draft PRs cannot be merged, so there is no need to worry about merging in bad code. Github has documentation about draft PRs [here.](#https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/about-pull-requests#draft-pull-requests)

Reviewer: ```suggestion
If you have an incomplete project you would like feedback on, you can make a draft PR. Prepend `[WIP]` to the PR title when making draft pull requests. Draft PRs cannot be merged, so there is no need to worry about merging in bad code. GitHub has documentation about draft PRs [here.](#https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/about-pull-requests#draft-pull-requests)
```

---
section: Guides
chapter: General
title: Git
description: Contributing to a NUbots project with git.
slug: /guides/general/git
---

## Prerequisites

This guide assumes you know a few basic shell commands, and the basics of git, such as adding, committing, and pushing to a repository.

## Getting Code Into a Repository

> Let's suppose you have created a file, you have made which you would like to add to a NUbots repository, called `my_file.cpp`, which is part of your project to make the robot stand up.

Reviewer 3: ```suggestion
Let's suppose you have created a file which you would like to add to a NUbots repository called `MakeRobotStand.cpp`, which is part of your project to make the robot stand up.
```
